pkg/partition/partition.go

```diff
@@ -0,0 +1,109 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package partition
+
+import (
+	"fmt"
+
+	"github.com/diskfs/go-diskfs"
+	"github.com/diskfs/go-diskfs/partition/gpt"
+	"k8s.io/klog"
+)
+
+const (
+	// BytesRequiredForGPTHeader is the total bytes required to store the GPT header
+	// 128 bytes are required per partition, total no of partition supported by GPT = 128
+	// Therefore, total bytes = 128*128
+	BytesRequiredForGPTHeader = 16384
+
+	// GPTPartitionStartByte is the byte on the disk at which the first partition starts.
+	// Normally partition starts at 1MiB, (as done by fdisk utility). This is done to
+	// align the partition start to physical block sizes on the disk.
+	GPTPartitionStartByte = 1048576
+)
+
+// Disk struct represents a disk which needs to be partitioned
+type Disk struct {
+	// DevPath is the /dev/sdX entry of the disk
+	DevPath string
+	// DiskSize is size of disk in bytes
+	DiskSize uint64
+	// LogicalBlockSize is the block size of the disk normally 512 or 4k
+	LogicalBlockSize uint64
+
+	table *gpt.Table
+}
+
+// ApplyPartitionTable applies the partition table to the given disk.
+func (d *Disk) ApplyPartitionTable() error {
+	if len(d.table.Partitions) == 0 {
+		return fmt.Errorf("no partitions specified in partition table")
+	}
+	diskFD, err := diskfs.Open(d.DevPath)
+	if err != nil {
+		return fmt.Errorf("error opening diskFD for write: %v", err)
+	}
+	err = diskFD.Partition(d.table)
+	if err != nil {
+		return fmt.Errorf("unable to create partition table. %v", err)
+	}
+	return nil
+}
+
+// CreatePartitionTable creates gpt partition table structure with protective MBR turned on
+func (d *Disk) CreatePartitionTable() error {
+	if d.DiskSize == 0 {
+		return fmt.Errorf("disk size is zero, unable to initialize partition table")
+	}
+	if d.LogicalBlockSize == 0 {
+		klog.Warningf("logical block size of %s not set, falling back to 512 bytes", d.DevPath)
+		klog.Warning("partitioning may fail.")
+		d.LogicalBlockSize = 512
+	}
+	// set protective MBR to true.
+	// https://en.wikipedia.org/wiki/GUID_Partition_Table#Protective_MBR_(LBA_0)
+	d.table = &gpt.Table{
+		LogicalSectorSize: int(d.LogicalBlockSize),
+		ProtectiveMBR:     true,
+	}
+	return nil
+}
+
+// AddPartition is used to add a partition to the partition table.
+// Currently only a single partition can be created. i.e The method can be called only once for a disk
+// TODO: @akhilerm, add method to create partiton with given size
+func (d *Disk) AddPartition() error {
+	var startSector, endSector uint64
+	if len(d.table.Partitions) == 0 {
+		// First sector of partition is aligned at 1MiB
+		startSector = (GPTPartitionStartByte) / d.LogicalBlockSize
```

Comment:
almost all the time GPTPartitionStartByte is divisible by LogicalBlockSize, however how about making it as:
if (GPTPartitionStartByte % d.LogicalBlockSize == 0)
startSector = GPTPartitionStartByte / d.LogicalBlockSize
else
startSector = GPTPartitionStartByte / d.LogicalBlockSize + 1

Reply:
@vishnuitta , the startbyte of partitions should be always aligned to logical blocks. Otherwise it will impact performance. (quoted from fdisk). This is the reason, the partition start is always set at 1MB, so that block sizes of 512, 4k, 16k or something that comes up later can also be handled without degrading performance.

Reply:
what if partitionStartByte is NOT divisible by LogicalBlockSize?

Reply:
partitionStartByte = 1MB = 1048576 bytes.

Hard drive vendors, have a fixed format for logical/physical sector sizes.  Current logical block sizes are 512 and 4k. 

@vishnuitta , are there cases of block sizes which are not multiples of 512 bytes. I am not sure of that.

References :
https://www.thomas-krenn.com/en/wiki/Advanced_Sector_Format_of_Block_Devices
[Western Digital](https://documents.westerndigital.com/content/dam/doc-library/en_us/assets/public/western-digital/collateral/white-paper/white-paper-advanced-format.pdf)
[Seagate] (https://www.seagate.com/in/en/tech-insights/advanced-format-4k-sector-hard-drives-master-ti/)
Logical Block size in [NVMe](https://metebalci.com/blog/a-quick-tour-of-nvm-express-nvme/#nvme-with-examples)

Reply:
may not be.. but, just considering a case..
I vaguely remember disks of physical sectors of 4096 + few bytes to store parity..